pkg/coscheduling/core/core.go

```diff
@@ -0,0 +1,267 @@
+	pgCopy.Status.Scheduled++
+
+	if pgCopy.Status.Scheduled >= pgCopy.Spec.MinMember {
+		pgCopy.Status.Phase = v1alpha1.PodGroupScheduled
+	} else {
+		pgCopy.Status.Phase = v1alpha1.PodGroupScheduling
+		if pgCopy.Status.ScheduleStartTime.IsZero() {
+			pgCopy.Status.ScheduleStartTime = metav1.Time{Time: time.Now()}
+		}
+	}
+	pg.Status = pgCopy.Status
+	if pgCopy.Status.Phase != pg.Status.Phase {
+
+		pg, err := pgMgr.pgLister.PodGroups(pgCopy.Namespace).Get(pgCopy.Name)
+		if err != nil {
+			klog.Error(err)
+			return
+		}
+		patch, err := util.CreateMergePatch(pg, pgCopy)
+		if err != nil {
+			klog.Error(err)
+			return
+		}
+		if err := pgMgr.PatchPodGroup(pg.Name, pg.Namespace, patch); err != nil {
+			klog.Error(err)
+			return
+		}
+		pg.Status.Phase = pgCopy.Status.Phase
+	}
+	pg.Status.Scheduled = pgCopy.Status.Scheduled
+	return
+}
+
+// getCreationTime return creation time of podgroup or pod
+func (pgMgr *PodGroupManager) GetCreationTime(pod *corev1.Pod, ts time.Time) time.Time {
+	pgName := util.GetPodGroupLabel(pod)
+	if len(pgName) == 0 {
+		return ts
+	}
+	pg, err := pgMgr.pgLister.PodGroups(pod.Namespace).Get(pgName)
+	if err != nil {
+		return ts
+	}
+	return pg.CreationTimestamp.Time
+}
+
+// AddToDenyCache add podGroup to the back list cache
+func (pgMgr *PodGroupManager) AddToDenyCache(pgFullName string) {
+	pgMgr.lastDeniedPG.Add(pgFullName, "", 20*time.Second)
+}
+
+// PatchPodGroup
+func (pgMgr *PodGroupManager) PatchPodGroup(pgName string, namespace string, patch []byte) error {
+	_, err := pgMgr.pgClient.SchedulingV1alpha1().PodGroups(namespace).Patch(context.TODO(), pgName,
+		types.MergePatchType, patch, metav1.PatchOptions{})
+	return err
+}
+
+// GetPodGroup get PodGroup from cache
+func (pgMgr *PodGroupManager) GetPodGroup(pod *corev1.Pod) (string, *v1alpha1.PodGroup) {
+	pgName := util.GetPodGroupLabel(pod)
+	if len(pgName) == 0 {
+		return "", nil
+	}
+	pg, err := pgMgr.pgLister.PodGroups(pod.Namespace).Get(pgName)
+	if err != nil {
+		return fmt.Sprintf("%v/%v", pod.Namespace, pgName), nil
+	}
+	return fmt.Sprintf("%v/%v", pod.Namespace, pgName), pg
+}
+
+// fillOccupiedObj fills the following objects to pod group cache
+// 1. Fill the pod of group, we may use it to compute resource
+// 2. Fill the MinResource if not set, according to pod resource requirement and minMember of a group
+// 3. Fill the field `OccupiedBy` according to owner reference, use the first one if a pod has multi
+func (pgMgr *PodGroupManager) fillOccupiedObj(pg *v1alpha1.PodGroup, pod *corev1.Pod) error {
+	if pg == nil {
+		return fmt.Errorf("PodGroup is nil")
+	}
+	pgFullName := GetNamespacedName(pg)
+
+	var refs []string
+	for _, ownerRef := range pod.OwnerReferences {
+		refs = append(refs, string(ownerRef.UID))
+	}
+	if pg.Status.OccupiedBy == "" {
+		// if do not have refs or initialize, return nil
+		if len(refs) != 0 {
+			// initialize pod group
+			sortkeys.Strings(refs)
+			pg.Status.OccupiedBy = strings.Join(refs, ",")
```

Comment:
I guess (not 100% sure though) client-go would reject this behavior - mutating an API object from sharedInformer.

If not, we should also avoid modifying a shared in-memory object as that serves as a shared singleton across all callers of the sharedInformer.

Reply:
I would like remove the fill logic!
Then open anther PR to move the logic to podGroup controller.

WDYT?

Reply:
I would like remove the fill logic!
Then open anther PR to move the logic to podGroup controller.

WDYT?

Reply:
SG.

Reply:
@Huang-Wei  Updated